extensions/src/audio_ports_activation.rs

```diff
@@ -0,0 +1,176 @@
+            is_input: bool,
+            port_index: u32,
+            is_active: bool,
+            sample_size: u32,
+        ) -> bool {
+            match plugin.use_extension(&self.0).set_active {
+                None => false,
+                Some(set_active) => {
+                    // SAFETY: This type ensures the function pointer is valid.
+                    unsafe {
+                        set_active(
+                            plugin.as_raw(),
+                            is_input,
+                            port_index,
+                            is_active,
+                            sample_size,
+                        )
+                    }
+                }
+            }
+        }
+    }
+}
+
+#[cfg(feature = "clack-plugin")]
+mod plugin {
+    use super::*;
+    use clack_plugin::extensions::prelude::*;
+
+    pub trait PluginAudioPortsActivationImpl {
+        fn can_activate_while_processing(&mut self) -> bool;
+    }
+
+    // NOTE: we have this trait split b/c if `PluginAudioPortsActivationImpl::can_activate_while_processing` is true,
+    // then `set_active` is called from the audio thread (otherwise it is called from the main thread).
+    pub trait PluginAudioPortsActivationSharedImpl {
+        fn set_active(
+            &self,
+            is_input: bool,
+            port_index: u32,
+            is_active: bool,
+            sample_size: SampleSize,
+        ) -> bool;
+    }
+
+    // SAFETY: The given struct is the CLAP extension struct for the matching side of this extension.
+    unsafe impl<P: Plugin> ExtensionImplementation<P> for PluginAudioPortsActivation
+    where
+        for<'a> P::Shared<'a>: PluginAudioPortsActivationSharedImpl,
+        for<'a> P::MainThread<'a>: PluginAudioPortsActivationImpl,
+    {
+        const IMPLEMENTATION: RawExtensionImplementation =
+            RawExtensionImplementation::new(&clap_plugin_audio_ports_activation {
+                can_activate_while_processing: Some(can_activate_while_processing::<P>),
+                set_active: Some(set_active::<P>),
+            });
+    }
+
+    #[allow(clippy::missing_safety_doc)]
+    unsafe extern "C" fn can_activate_while_processing<P: Plugin>(
+        plugin: *const clap_plugin,
+    ) -> bool
+    where
+        for<'a> P::MainThread<'a>: PluginAudioPortsActivationImpl,
+    {
+        PluginWrapper::<P>::handle(plugin, |plugin| {
+            Ok(plugin
+                .main_thread()
+                .as_mut()
+                .can_activate_while_processing())
+        })
+        .unwrap_or(false)
+    }
+
+    #[allow(clippy::missing_safety_doc)]
+    unsafe extern "C" fn set_active<P: Plugin>(
+        plugin: *const clap_plugin,
+        is_input: bool,
+        port_index: u32,
+        is_active: bool,
+        sample_size: u32,
+    ) -> bool
+    where
+        for<'a> P::Shared<'a>: PluginAudioPortsActivationSharedImpl,
+    {
+        let sample_size = SampleSize::from_raw(sample_size).unwrap();
```

Comment:
You want this line to be inside the `handle` callback: panicking in a C callback called via FFI is Undefined Behavior and the host won't be able to handle it.

The `PluginWrapper::handle` function wraps the given callback in `catch_unwind` and automatically logs everything if a panic happens, and it just returns `None`. 🙂

Reply:
:pray: